### PR TITLE
Untyped Editors within override service

### PR DESCRIPTION
**Pull request by lramos15**
This changes the override service to only work with untyped editor inputs. Additionally it adds support for the untitled case where we ask the registered editor to create an input and they must decide how the resource is created.

### CODE REVIEW

File: src/vs/workbench/contrib/preferences/common/preferencesContribution.ts
Comment (bpasero) [inline src/vs/workbench/contrib/preferences/common/preferencesContribution.ts:71]: Nicer: `({resource, options}, group): IEditorInputWithOptions => {`

File: src/vs/workbench/contrib/terminal/browser/terminalService.ts
Comment (bpasero) [inline src/vs/workbench/contrib/terminal/browser/terminalService.ts:172]: Nicer: `({resource, options}, group): IEditorInputWithOptions => {`

File: src/vs/workbench/contrib/notebook/browser/notebookServiceImpl.ts
Comment (bpasero) [inline src/vs/workbench/contrib/notebook/browser/notebookServiceImpl.ts:157]: Nicer: `({resource, options}, group): IEditorInputWithOptions => {`

File: src/vs/workbench/contrib/customEditor/browser/customEditors.ts
Comment (bpasero) [inline src/vs/workbench/contrib/customEditor/browser/customEditors.ts:127]: Why not `editorInput.resource` here?

**Review by bpasero — CHANGES_REQUESTED** [review]
Gave it a quick review but didn't drill into the changes in the `editorOverrideService` since I consider you know well what you are doing there.

I am not 100% sure we do the right thing with regards to:

```ts
resource = URI.from({ scheme: Schemas.untitled });
```

When no resource is defined. I understand that we use it to resolve the override, but imho it should not be passed as untyped input to the factory functions (maybe we don't, I didn't check that). In other words, the original untyped input should be preserved when passed to the factory.

**Comment by lramos15** [comment]
> Gave it a quick review but didn't drill into the changes in the `editorOverrideService` since I consider you know well what you are doing there.
> 
> I am not 100% sure we do the right thing with regards to:
> 
> ```ts
> resource = URI.from({ scheme: Schemas.untitled });
> ```
> 
> When no resource is defined. I understand that we use it to resolve the override, but imho it should not be passed as untyped input to the factory functions (maybe we don't, I didn't check that). In other words, the original untyped input should be preserved when passed to the factory.

I'm not sure I understand. We just pass the untyped editor input through which contains the necessary info for the consumers to generate the resource. It is undefined in the non associative case and contains the resource in the associative case.

**Comment by bpasero** [comment]
> I'm not sure I understand. We just pass the untyped editor input through which contains the necessary info for the consumers to generate the resource. It is undefined in the non associative case and contains the resource in the associative ca

Yeah and that is good. But truth is that you "invent" a `resource` if none is defined here:

https://github.com/microsoft/vscode/blob/c772c9b1a2c60ae6b2c472b8ea874bb169001e3d/src/vs/workbench/services/editor/browser/editorOverrideService.ts#L92-L92

Then, the `resource` is used in lots of places (e.g. to figure out if the active editor matches):

https://github.com/microsoft/vscode/blob/c772c9b1a2c60ae6b2c472b8ea874bb169001e3d/src/vs/workbench/services/editor/browser/editorOverrideService.ts#L125-L125

I think this is dangerous. I would have expected that the `resource` is only used to figure out the override, but otherwise remains `undefined`. In other words: we should only look at the incoming `resource` if it was not defined by the editor override service internally, no?

**Comment by lramos15** [comment]
> Or was this change done for convenience to not have resource: URI | undefined?

Right. In the non associative case we make up the fake resource `untitled://` and use that as we can guarantee the future resource is going to be `untitled://**`. This allows me to avoid writing a lot of redundant code such as `if (resource.scheme === Schemas.untitled || resource === undefined)` and instead I can just treat both associative and non associative untitled resources as the same. The places were I use resource this should be generally safe in, even in the active editor case which will always return false as you will never have an active editor with just an untitled scheme with no authority or path. This same logic would hold if we left resource undefined.

**Comment by bpasero** [comment]
OK, just wanted to make sure it doesn't cause any issues.

**Comment by lramos15** [comment]
Right now I do some patching to make sure things work with typed editor inputs as well. The next goal is to change the signature of `openEditor/replaceEditor/openEditors` to only use untyped editor inputs. I think this PR is ready to go though for now and that would be another one.

File: src/vs/workbench/test/browser/workbenchTestServices.ts
Comment (bpasero) [inline src/vs/workbench/test/browser/workbenchTestServices.ts:1483]: Weird whitespace used here

File: src/vs/workbench/common/editor/sideBySideEditorInput.ts
Comment (bpasero) [inline src/vs/workbench/common/editor/sideBySideEditorInput.ts:146]: This seems weird to me, I do not think we support such a untyped side by side editor input? The closest is this:

https://github.com/microsoft/vscode/blob/97d770d4edf9d8ed44d02ab2b1cef284e6d1dc29/src/vs/workbench/common/editor/diffEditorInput.ts#L140-L154

File: src/vs/workbench/services/editor/browser/editorService.ts
Comment (bpasero) [inline src/vs/workbench/services/editor/browser/editorService.ts:584]: Some empty lines that can be collapsed

File: src/vs/workbench/services/editor/browser/editorService.ts
Comment (bpasero) [inline src/vs/workbench/services/editor/browser/editorService.ts:581]: I think we should start to introduce a `isTypedEditorInput` that asserts that the thing is `IEditorInput` to prevent the explicit casts. I could use it in more places where I have a similar pattern...

File: src/vs/workbench/services/editor/browser/editorService.ts
Comment (bpasero) [inline src/vs/workbench/services/editor/browser/editorService.ts:695]: I prefer statements without newline within `for` unless you have a comment above

**Review by bpasero — CHANGES_REQUESTED** [review]
Some first feedback, I unfortunately run out of time to look into the changes in editor service which seem rather large and involved, maybe we do a call tomorrow and you walk me through it?

File: src/vs/workbench/services/editor/browser/editorService.ts
Comment (bpasero) [reply src/vs/workbench/services/editor/browser/editorService.ts:581]: I pushed to `main` a new `isEditorInput` method that can be used.

File: src/vs/workbench/services/editor/browser/editorService.ts
Comment (bpasero) [inline src/vs/workbench/services/editor/browser/editorService.ts:944]: This is dangerous, no? Because we have a `EditorInput` in hand but you only take it's `resource` going forward. What about `SideBySideEditorInput` and `DiffEditorInput`? Because in the `resolveEditorOverride` method we seem to be supporting side by side?

https://github.com/microsoft/vscode/blob/48cbb9430a7d6089d5e6ddb52cc5e1213b317cba/src/vs/workbench/services/editor/browser/editorOverrideService.ts#L83-L83

File: src/vs/workbench/services/editor/browser/editorService.ts
Comment (bpasero) [inline src/vs/workbench/services/editor/browser/editorService.ts:721]: This seems like an unnecessary variable to add.

File: src/vs/workbench/services/editor/browser/editorService.ts
Comment (bpasero) [inline src/vs/workbench/services/editor/browser/editorService.ts:723]: This code is quite involved. I would suggest to make `resolveEditorOverride` accept `IEditorInputWithOptions` to make this easier to read and deprecate that overload so that we know we have to remove it eventually?

File: src/vs/workbench/services/editor/browser/editorService.ts
Comment (bpasero) [inline src/vs/workbench/services/editor/browser/editorService.ts:741]:
```suggestion
					targetGroupEditors.push(editorOverride);
```

File: src/vs/workbench/services/editor/browser/editorService.ts
Comment (bpasero) [inline src/vs/workbench/services/editor/browser/editorService.ts:540]: Should this maybe be renamed to `resolveGroupAndOptions`?

File: src/vs/workbench/services/editor/browser/editorService.ts
Comment (bpasero) [inline src/vs/workbench/services/editor/browser/editorService.ts:580]:
```suggestion
			if (!isEditorInput(editor)) {
```

File: src/vs/workbench/services/editor/browser/editorService.ts
Comment (bpasero) [inline src/vs/workbench/services/editor/browser/editorService.ts:581]: Isn't this replacing all of the `options` that were maybe passed in? Why not mixing them in?

Comment (bpasero) [reply src/vs/workbench/services/editor/browser/editorService.ts:581]: Oh wow, this is confusing, the `options` are from the untyped input. Then the code is not doing anything right?

File: src/vs/workbench/services/editor/browser/editorService.ts
Comment (bpasero) [inline src/vs/workbench/services/editor/browser/editorService.ts:521]: This is weird. `resolveEditorOverride` only seems to accept `IUntypedEditorInput` but you are passing in `EditorInput`.

**Review by bpasero — CHANGES_REQUESTED** [review]
In this second review I focused on `editorService.ts` changes.

**Comment by bpasero** [comment]
@lramos15 as for the test failures, I think the issue is that tests open editors, e.g. here:

https://github.com/microsoft/vscode/blob/27c543871c4507c1fe61cf101d0bee7b204bd331/src/vs/workbench/services/workingCopy/test/browser/workingCopyBackupTracker.test.ts#L359-L359

But then get overridden to a different editor input. As such, we fail to recognize the editor at a later point:

https://github.com/microsoft/vscode/blob/27c543871c4507c1fe61cf101d0bee7b204bd331/src/vs/workbench/services/workingCopy/test/browser/workingCopyBackupTracker.test.ts#L385-L385

I am not sure why the override happens, but probably because of our default override that always kicks in and returns text editors.

If you pass in `EditorOverride.DISABLED` in those tests that open editors, it should probably work. But for all other tests that open test inputs, they are very likely now not opening that input but a default text input based on the default override, so maybe it needs a general pass to figure out we are still testing the right thing?

File: src/vs/workbench/common/editor/sideBySideEditorInput.ts
Comment (bpasero) [reply src/vs/workbench/common/editor/sideBySideEditorInput.ts:146]: If so, should the `diffEditorInput` call `super.toUntyped`?

Comment (lramos15) [reply src/vs/workbench/common/editor/sideBySideEditorInput.ts:146]: Or should we create an untyped side by side input?

File: src/vs/workbench/services/editor/browser/editorService.ts
Comment (lramos15) [reply src/vs/workbench/services/editor/browser/editorService.ts]: The reason I do this is just to get the added activation property.

File: src/vs/workbench/common/editor/sideBySideEditorInput.ts
Comment (bpasero) [reply src/vs/workbench/common/editor/sideBySideEditorInput.ts:146]: There is too many things going on here at once, I will remove this overload and follow up in https://github.com/microsoft/vscode/issues/127131

**Comment by bpasero** [comment]
A log of things I did today or figured out:

* [ ] **Untyped side by side editors**

I think we should decouple the support out of this PR and I filed https://github.com/microsoft/vscode/issues/127131 and removed the method `toUntyped` from the side by side editor input.

via https://github.com/microsoft/vscode/pull/126918/commits/3b5f243562c91eace680bfd54179a1b9113ec6c3

* [ ] **Disabled untitled notebook test**

Given the test is commented out we really need to ensure that notebooks provide the `createUntitledEditorInput` factory or provide a fallback to not break existing scenarios.

* [ ] **`IEditorService#replaceEditors**

Sorry for the rewrite but I am having a hard time to understand the very condense format you ended up with. Besides, I didn't like the fact that `IResourceEditorReplacement` suddenly has a top level `options` property. It should really only be the `replacement.options` property to be used because untyped editor inputs always carry options around. Wouldn't it be weird if someone defines options both top level and in `replacement`? Which one would win?

Besides, I figured out that the `editor` in the replacement that is to be replaced should never be untyped but always typed because the intent is to replace an existing editor, so I filed https://github.com/microsoft/vscode/issues/127134

via https://github.com/microsoft/vscode/pull/126918/commits/4839564e70598d5adff929b7f5453b63beb5111c

* [ ] **`IEditorService#resolveGroupAndOptions`**

Given we no longer return a typed editor input from this method, lots of cleanup was possible. For one, it would never return `undefined` anymore. And the fact that it returned editor options was imho confusing because the only use case for that was to return the `EditorGroupActivation` strategy. So I simplified that and renamed to `findTargetGroup`.

via https://github.com/microsoft/vscode/pull/126918/commits/192001423906ceb963bd82f54fec0b4705094e8e

* [ ] **`doResolveEditorOverride`**

This needs revisit. First of all I think it should really not live in editor service but in override service. Then I think it does some wonky things with regards to resolving the editor group, when it was resolved before?

I did not review the method yet because I think it shouldn't exist like that. Also, what is `populateEditorId`?

Related https://github.com/microsoft/vscode/issues/127137

* [ ] **`openEditors`**

I think this had a bug actually where you would ignore the editor if `toUntyped` would return `undefined`. My changes try to implement a similar strategy as we have elsewhere:
* if editor does not support untyped, take it as is
* if override is `None`, take it as is
* if override is `Abort`, skip it
* otherwise, take the override

via https://github.com/microsoft/vscode/pull/126918/commits/1d84aa00085c33a013326191bb63640516643c8c

* [ ] **`IEditorOverrideService`**

I was not able to review the changes in there, but I also feel you own this piece entirely as I never really reviewed code in there.

File: src/vs/workbench/services/editor/browser/editorService.ts
Comment (bpasero) [reply src/vs/workbench/services/editor/browser/editorService.ts]: This no longer applies.

File: src/vs/workbench/services/editor/browser/editorService.ts
Comment (bpasero) [reply src/vs/workbench/services/editor/browser/editorService.ts]: I think this no longer applies

File: src/vs/workbench/services/editor/browser/editorService.ts
Comment (bpasero) [reply src/vs/workbench/services/editor/browser/editorService.ts]: I think this is ok now, eventully typed editor support will be gone from the editor service methods.

**Comment by bpasero** [comment]
**From our meeting:**

**P1 (should do before merge) **
* implement `createUntitledEditorInput` for notebooks @lramos15 
* implement `createUntitledEditorInput` for custom editors @lramos15 
* check if side by side editor input is broken now or whether it needs `toUntyped` @lramos15 (follow up with Pung)
* change `doResolveEditorOverride` (rename to something else? `doResolveEditor` to return `group` and `options` as we had it before @lramos15 
* use `doResolveEditorOverride` in `openEditors` @lramos15 

**P2 (can do after merge) **
* change `populateEditorId` to `resolveEditorId` and do the mixin from the calling side @lramos15 
* cleanup `replaceEditors` to not pass untyped as `editor` (https://github.com/microsoft/vscode/issues/127134) @bpasero 
* use more `IEditorWithOptions` in editor service methods @bpasero

**Comment by bpasero** [comment]
Looks decent to me, is still super complex code and very hard for me to figure out if this causes regressions or not so please make sure to have this covered in test plan for the cases touched:
* `openEditor`
* `openEditors`
* `replaceEditors`

Pushed https://github.com/microsoft/vscode/commit/ee290519377dbe3e4deb3b325269297b3950030d as follow up :lipstick:

**Pull request merged**
